pylint/testutils/functional/test_file.py

```diff
@@ -0,0 +1,109 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+
+import configparser
+import sys
+from os.path import basename, exists, join
+from typing import List, Tuple
+
+
+def parse_python_version(ver_str: str) -> Tuple[int, ...]:
+    """Convert python version to a tuple of integers for easy comparison."""
+    return tuple(int(digit) for digit in ver_str.split("."))
+
+
+class NoFileError(Exception):
+    pass
+
+
+if sys.version_info >= (3, 8):
+    from typing import TypedDict
+else:
+    from typing_extensions import TypedDict
+
+
+class TestFileOptions(TypedDict):
+    min_pyver: Tuple[int, ...]
+    max_pyver: Tuple[int, ...]
+    min_pyver_end_position: Tuple[int, ...]
+    requires: List[str]
+    except_implementations: str  # Type is actually comma separated list of string
+    exclude_platforms: str  # Type is actually comma separated list of string
+
+
+# mypy need something literal, we can't create this dynamically from TestFileOptions
+POSSIBLE_TEST_OPTIONS = {
+    "min_pyver",
+    "max_pyver",
+    "min_pyver_end_position",
+    "requires",
+    "except_implementations",
+    "exclude_platforms",
+    "exclude_platforms",
+}
+
+
+class FunctionalTestFile:
+    """A single functional test case file with options."""
+
+    _CONVERTERS = {
+        "min_pyver": parse_python_version,
+        "max_pyver": parse_python_version,
+        "min_pyver_end_position": parse_python_version,
+        "requires": lambda s: s.split(","),
+    }
+
+    def __init__(self, directory: str, filename: str) -> None:
+        self._directory = directory
+        self.base = filename.replace(".py", "")
+        self.options: TestFileOptions = {
+            "min_pyver": (2, 5),
+            "max_pyver": (4, 0),
+            "min_pyver_end_position": (3, 8),
+            "requires": [],
+            "except_implementations": "",
+            "exclude_platforms": "",
+        }
+        self._parse_options()
+
+    def __repr__(self) -> str:
+        return f"FunctionalTest:{self.base}"
+
+    def _parse_options(self) -> None:
+        cp = configparser.ConfigParser()
+        cp.add_section("testoptions")
+        try:
+            cp.read(self.option_file)
+        except NoFileError:
+            pass
+
+        for name, value in cp.items("testoptions"):
+            conv = self._CONVERTERS.get(name, lambda v: v)
+
+            assert (
+                name in POSSIBLE_TEST_OPTIONS
```

Comment:
`TestFileOptions.__annotations__` would also work, but I'm not sure if you're supposed to access it..

Reply:
I don't like typed dict, it's only marginally better than a dict. Might refactor that class later.